core/src/main/java/org/elasticsearch/index/query/QueryFilterBuilder.java

```diff
@@ -78,6 +97,58 @@ protected void buildFQuery(XContentBuilder builder, Params params) throws IOExce
     }
 
     @Override
+    public Query toQuery(QueryParseContext parseContext) throws QueryParsingException, IOException {
+        // inner query builder can potentially be `null`, in that case we ignore it
+        if (this.queryBuilder == null) {
+            return null;
+        }
+        Query innerQuery = this.queryBuilder.toQuery(parseContext);
+        if (innerQuery == null) {
+            return null;
+        }
+        Query query = new ConstantScoreQuery(innerQuery);
+        if (queryName != null) {
+            parseContext.addNamedQuery(queryName, query);
+        }
+        return query;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(queryBuilder, queryName);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) {
+            return true;
+        }
+        if (obj == null || getClass() != obj.getClass()) {
+            return false;
+        }
+        QueryFilterBuilder other = (QueryFilterBuilder) obj;
+        return Objects.equals(queryBuilder, other.queryBuilder) &&
+                Objects.equals(queryName, other.queryName);
+    }
+
+    @Override
+    public QueryFilterBuilder readFrom(StreamInput in) throws IOException {
+        QueryBuilder innerQueryBuilder = in.readNamedWriteable();
+        String queryName = in.readOptionalString();
+        if (queryName == null) {
+            return new QueryFilterBuilder(innerQueryBuilder);
+        } else {
+            return new FQueryFilterBuilder(innerQueryBuilder).queryName(queryName);
```

Comment:
What if we accept a little bit of code duplication and separate FQueryFilterBuilder and QueryFilterBuilder in so far as they both extend QueryBuilder. Then FQueryBuilder would be the one that has a `queryName`. The only thing that remains slightly odd then is that when rendering out, FQuery still uses the QueryFilterBuilder name ("query") because it essentially wraps it. But thats the state of the DSL at the moment.